File: 99/scala/p05.scala
/**
 * P05: Reverse a list
 */

package s99.lists

package object p05 {
  def reverse(ls: List[_]): List[_] = ls match {
    case Nil => Nil
    case x :: xs => reverse(xs) ::: List(x)
  }
  
  // built-in
  def reverse2(ls: List[_]) = ls.reverse
  
  // another
  def reverse3[T](ls: List[T]): List[T] = (ls foldLeft List[T]()) { (acc, e) => e :: acc }
}

File: 99/ruby/p05.rb
# P05: Reverse a list

require_relative "lists.rb"

# will just monkey patch list class
class List
  # by going to end and constructing in reverse order
  def self.reverse(ls)
    reversed = Nil.instance
    ls.each do |el|
      reversed = reversed << el
    end
    reversed
  end

  # in somewhat recursive manner (worse that first)
  def self.reverse2(ls)
    if ls.empty?
      ls
    else
      self.reverse2(ls.tail) >> ls.head
    end
  end
end

#test
class Tests
  def self.reverse
    ls = List["one", "two", "three", "four", "five"]
    
    puts ls
    puts List.reverse ls
    puts List.reverse2 ls
  end
end